src/qibo/transpiler/router.py

```diff
@@ -230,130 +251,80 @@ def routed_circuit(self, circuit_kwargs: Optional[dict] = None):
         return self._routed_blocks.circuit(circuit_kwargs=circuit_kwargs)
 
     def final_layout(self):
-        """Returns the final physical-circuit qubits mapping."""
-        unsorted_dict = {
-            "q" + str(self.circuit_to_physical(i)): i
-            for i in range(len(self._circuit_logical))
-        }
+        """Returns the final physical-logical qubits mapping."""
+
+        # 1# return {"q0": lq_num0, "q1": lq_num1, ...}
+        unsorted_dict = {"q" + str(i): self._p2l[i] for i in range(self._nqubits)}
 
         return dict(sorted(unsorted_dict.items()))
 
-    def update(self, swap: tuple):
-        """Updates the logical-physical qubit mapping after applying a ``SWAP``
+    def update(self, swap_l: tuple):
+        """Updates the qubit mapping after applying a ``SWAP``
 
         Adds the :class:`qibo.gates.gates.SWAP` gate to the routed blocks.
         Method works in-place.
 
         Args:
             swap (tuple): tuple containing the logical qubits to be swapped.
         """
-        physical_swap = self.logical_to_physical(swap, index=True)
-        self._routed_blocks.add_block(
-            Block(qubits=physical_swap, gates=[gates.SWAP(*physical_swap)])
-        )
-        self._swaps += 1
-        idx_0, idx_1 = self._circuit_logical.index(
-            swap[0]
-        ), self._circuit_logical.index(swap[1])
-        self._circuit_logical[idx_0], self._circuit_logical[idx_1] = swap[1], swap[0]
+
+        swap_p = self.logical_pair_to_physical(swap_l)
+
+        # 2# add the real SWAP gate, not a temporary circuit
+        if not self._temporary:
+            self._routed_blocks.add_block(
+                Block(qubits=swap_p, gates=[gates.SWAP(*swap_p)])
+            )
+            self._swaps += 1
+
+        # 1# update the bidirectional mapping
+        p1, p2 = swap_p
+        l1, l2 = swap_l
+        self._p2l[p1], self._p2l[p2] = l2, l1
+        self._l2p[l1], self._l2p[l2] = p2, p1
```

Comment:
It might be worth making this a small helper function, e.g. `def _update_mappings(self, logical_pair, physical_pair)`